File: Telemedicine-and-Doctor-Consultation-Platform-main/backend/routes/consultations.js
const router = require('express').Router();
const auth   = require('../middleware/auth');
const db     = require('../config/db');

// GET /api/consultations
router.get('/', auth, async (req, res) => {
  try {
    const { status, doctorId, patientId, date } = req.query;
    let query = 'SELECT * FROM consultations WHERE 1=1';
    let params = [];
    
    if (status)    { query += ' AND status = ?'; params.push(status); }
    if (doctorId)  { query += ' AND doctor_id = ?'; params.push(doctorId); }
    if (patientId) { query += ' AND patient_id = ?'; params.push(patientId); }
    if (date)      { query += ' AND scheduled_date = ?'; params.push(date); }
    
    const [rows] = await db.query(query, params);
    
    const formatted = rows.map(r => ({
      ...r,
      patientId: r.patient_id,
      doctorId: r.doctor_id,
      date: typeof r.scheduled_date === 'string' ? r.scheduled_date : r.scheduled_date.toISOString().slice(0, 10),
      time: r.scheduled_time
    }));
    
    res.json({ count: formatted.length, data: formatted });
  } catch (err) {
    res.status(500).json({ error: 'Database error', message: err.message });
  }
});

// GET /api/consultations/:id
router.get('/:id', auth, async (req, res) => {
  try {
    const [rows] = await db.query('SELECT * FROM consultations WHERE id = ?', [req.params.id]);
    if (!rows.length) return res.status(404).json({ error: 'Consultation not found' });
    
    const r = rows[0];
    res.json({
        ...r,
        patientId: r.patient_id,
        doctorId: r.doctor_id,
        date: typeof r.scheduled_date === 'string' ? r.scheduled_date : r.scheduled_date.toISOString().slice(0, 10),
        time: r.scheduled_time
    });
  } catch (err) {
    res.status(500).json({ error: 'Database error', message: err.message });
  }
});

// POST /api/consultations — book
router.post('/', auth, async (req, res) => {
  try {
    const { patientId, doctorId, type, date, time, spec, reason } = req.body;
    if (!patientId || !doctorId) return res.status(400).json({ error: 'patientId and doctorId required' });
    
    const [countRows] = await db.query('SELECT COUNT(*) as count FROM consultations');
    const newId = `C${String(countRows[0].count + 1).padStart(3,'0')}`;
    
    const dDate = date || new Date().toISOString().slice(0,10);
    const dTime = time || '10:00:00';
    
    await db.query(`
      INSERT INTO consultations 
      (id, patient_id, doctor_id, type, speciality, scheduled_date, scheduled_time, status, reason) 
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
    `, [newId, patientId, doctorId, type || 'video', spec || 'General', dDate, dTime, 'scheduled', reason || '']);
    
    const [newRows] = await db.query('SELECT * FROM consultations WHERE id = ?', [newId]);
    const r = newRows[0];
    res.status(201).json({
        ...r, patientId: r.patient_id, doctorId: r.doctor_id, 
        date: typeof r.scheduled_date === 'string' ? r.scheduled_date : r.scheduled_date.toISOString().slice(0, 10),
        time: r.scheduled_time
    });
  } catch (err) {
    res.status(500).json({ error: 'Database error', message: err.message });
  }
});

// PATCH /api/consultations/:id/status
router.patch('/:id/status', auth, async (req, res) => {
  try {
    const { status, duration } = req.body;
    const updates = [];
    const params = [];
    
    if (status) { updates.push('status = ?'); params.push(status); }
    if (duration !== undefined) { updates.push('duration_min = ?'); params.push(duration); }
    
    if (!updates.length) return res.status(400).json({ error: 'Nothing to update' });
    
    params.push(req.params.id);
    await db.query(`UPDATE consultations SET ${updates.join(', ')} WHERE id = ?`, params);
    
    const [rows] = await db.query('SELECT * FROM consultations WHERE id = ?', [req.params.id]);
    if (!rows.length) return res.status(404).json({ error: 'Not found' });
    
    res.json(rows[0]);
  } catch (err) {
    res.status(500).json({ error: 'Database error', message: err.message });
  }
});

// GET /api/consultations/stats/today
router.get('/stats/today', auth, async (req, res) => {
  try {
    const today = new Date().toISOString().slice(0,10);
    const [rows] = await db.query('SELECT status, COUNT(*) as count FROM consultations WHERE scheduled_date = ? GROUP BY status', [today]);
    
    let stats = { total: 0, live: 0, waiting: 0, completed: 0 };
    rows.forEach(r => {
      stats.total += r.count;
      if (stats[r.status] !== undefined) stats[r.status] += r.count;
    });
    
    res.json(stats);
  } catch (err) {
    res.status(500).json({ error: 'Database error', message: err.message });
  }
});

module.exports = router;

File: Telemedicine-and-Doctor-Consultation-Platform-main/backend/routes/consultations.py
from flask import Blueprint, request, jsonify
from config import get_db_connection
from middleware.auth import token_required
from datetime import datetime

bp = Blueprint('consultations', __name__, url_prefix='/api/consultations')

def format_consultation(r):
    """Format consultation row"""
    date = r.get('scheduled_date')
    if isinstance(date, str):
        date_str = date
    else:
        date_str = date.isoformat()[:10] if date else None
    
    return {
        **r,
        'patientId': r.get('patient_id'),
        'doctorId': r.get('doctor_id'),
        'date': date_str,
        'time': r.get('scheduled_time')
    }

@bp.route('/', methods=['GET'])
@token_required
def get_consultations():
    try:
        status = request.args.get('status')
        doctor_id = request.args.get('doctorId')
        patient_id = request.args.get('patientId')
        date = request.args.get('date')
        
        query = 'SELECT * FROM consultations WHERE 1=1'
        params = []
        
        if status:
            query += ' AND status = %s'
            params.append(status)
        if doctor_id:
            query += ' AND doctor_id = %s'
            params.append(doctor_id)
        if patient_id:
            query += ' AND patient_id = %s'
            params.append(patient_id)
        if date:
            query += ' AND scheduled_date = %s'
            params.append(date)
        
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute(query, params)
        rows = cursor.fetchall()
        cursor.close()
        conn.close()
        
        formatted = [format_consultation(r) for r in rows]
        return jsonify({'count': len(formatted), 'data': formatted}), 200
    
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500

@bp.route('/<id>', methods=['GET'])
@token_required
def get_consultation(id):
    try:
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute('SELECT * FROM consultations WHERE id = %s', (id,))
        consultation = cursor.fetchone()
        cursor.close()
        conn.close()
        
        if not consultation:
            return jsonify({'error': 'Consultation not found'}), 404
        
        return jsonify(format_consultation(consultation)), 200
    
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500

@bp.route('/', methods=['POST'])
@token_required
def create_consultation():
    try:
        data = request.get_json()
        patient_id = data.get('patientId')
        doctor_id = data.get('doctorId')
        cons_type = data.get('type')
        date_str = data.get('date')
        time_str = data.get('time')
        spec = data.get('spec')
        reason = data.get('reason')
        
        if not patient_id or not doctor_id:
            return jsonify({'error': 'patientId and doctorId required'}), 400
        
        conn = get_db_connection()
        cursor = conn.cursor()
        
        # Get count for ID
        cursor.execute('SELECT COUNT(*) as count FROM consultations')
        result = cursor.fetchone()
        count = result['count'] + 1
        new_id = f'C{str(count).zfill(3)}'
        
        # Use defaults if not provided
        default_date = datetime.utcnow().date().isoformat()
        default_time = '10:00:00'
        
        # Insert consultation
        cursor.execute(
            '''INSERT INTO consultations 
               (id, patient_id, doctor_id, type, speciality, scheduled_date, scheduled_time, status, reason) 
               VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)''',
            (new_id, patient_id, doctor_id, cons_type or 'video', spec or 'General', 
             date_str or default_date, time_str or default_time, 'scheduled', reason or '')
        )
        conn.commit()
        
        # Get the created consultation
        cursor.execute('SELECT * FROM consultations WHERE id = %s', (new_id,))
        consultation = cursor.fetchone()
        cursor.close()
        conn.close()
        
        return jsonify(format_consultation(consultation)), 201
    
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500

@bp.route('/<id>/status', methods=['PATCH'])
@token_required
def update_consultation_status(id):
    try:
        data = request.get_json()
        status = data.get('status')
        duration = data.get('duration')
        
        updates = []
        params = []
        
        if status:
            updates.append('status = %s')
            params.append(status)
        if duration is not None:
            updates.append('duration_min = %s')
            params.append(duration)
        
        if not updates:
            return jsonify({'error': 'Nothing to update'}), 400
        
        params.append(id)
        
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute(f"UPDATE consultations SET {', '.join(updates)} WHERE id = %s", params)
        conn.commit()
        
        cursor.execute('SELECT * FROM consultations WHERE id = %s', (id,))
        consultation = cursor.fetchone()
        cursor.close()
        conn.close()
        
        if not consultation:
            return jsonify({'error': 'Not found'}), 404
        
        return jsonify(format_consultation(consultation)), 200
    
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500

@bp.route('/stats/today', methods=['GET'])
@token_required
def get_today_stats():
    try:
        today = datetime.utcnow().date().isoformat()
        
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute(
            'SELECT status, COUNT(*) as count FROM consultations WHERE scheduled_date = %s GROUP BY status',
            (today,)
        )
        rows = cursor.fetchall()
        cursor.close()
        conn.close()
        
        stats = {'total': 0, 'live': 0, 'waiting': 0, 'completed': 0}
        
        for row in rows:
            status_key = row['status'].lower()
            count = row['count']
            stats['total'] += count
            if status_key in stats:
                stats[status_key] = count
        
        return jsonify(stats), 200
    
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500
